### PR TITLE
Add missing type declarations and fix incorrect PHPDoc types

### PR DESCRIPTION
**Pull request by theodorejb**
This avoids the need for manual validation in several places, and makes it possible to use the library in codebases with static analysis, without having to suppress errors.

Also upgraded to PHPUnit 9, and added missing ext-json requirement.

### CODE REVIEW

**Comment by theodorejb** [comment]
@AaronAtDuo Can you review this please?

**Comment by mbish** [comment]
Thanks for adding these type annotations and fixing some typing inconsistencies. I don't think adding type information for static analysis tools removes the need for runtime type validation during the `Client` constructor. Since this information will likely be originating from some external data source (config file, database, etc.) enforcing that the data is in the shape we expect allows for better error reporting. PHP type checking is not enforced at runtime and therefore 

Similarly `exchangeAuthorizationCodeFor2FAResult` the data that's being validated originates from an API request and should be validated before use. 

It's harder to make an argument for `createAuthUrl` but in general I prefer the verbose, runtime checking in the handful of places that we have it.

If you could add the removed validation code and associated tests back in this PR looks good to me.

Thanks again for adding the typing information to this library. It greatly benefits from the consistency and clarity that automatically checked annotations can provide.

**Comment by theodorejb** [comment]
> PHP type checking is not enforced at runtime

@mbish You are mistaken. PHP type declarations are enforced at runtime, and will throw a `TypeError` if a value incompatible with the type is passed ([example](https://3v4l.org/ib9KV)). When the `strict_types` directive is enabled (as it is in `ClientTest.php`), any invalid/non-matching type will throw a `TypeError` at runtime.

I could add back tests which demonstrate this, but it feels somewhat superfluous to just be testing PHP's built-in validation of the declared types. Let me know if you still want such tests to be added.

**Comment by mbish** [comment]
I see, strict typing is optionally enforced. Since responsibility for enforcing type checking is opt-in and placed on the user of the library (i.e there's nothing we can do to ensure that the annotations in this library are respected) I'd prefer if the validation code was left in. We want to provide good defaults for users.

You raise a good point about the test code with the `strict_types` directive and the newly added annotations the tests aren't actually exercising the code they were written to exercise. Turning off that directive for `clientTest.php` makes sense in order to actually exercise the validation code. Moving the validation testing code into its own test file with `strict_types` disabled and keeping strict typing in `clientTest.php` would be even better but is a larger refactor.

**Comment by theodorejb** [comment]
> I see, strict typing is optionally enforced. Since responsibility for enforcing type checking is opt-in and placed on the user of the library (i.e there's nothing we can do to ensure that the annotations in this library are respected)

@mbish While it's true that *strict* type checks are optional, type checking itself is not opt-in or a burden placed on the user. Even without strict type checks enabled, PHP enforces that functions receive the declared types, and will throw a `TypeError` if an incompatible value is passed. The only difference is that without `strict_types` enabled, PHP is more forgiving in that it will coerce compatible values to the declared type. For example:

```php
function takesInt(int $val) {
    var_dump($val);
}

takesInt("5");   // int(5)
takesInt("foo"); // TypeError: takesInt(): Argument #1 ($val) must be of type int, string given
takesInt("");    // TypeError: takesInt(): Argument #1 ($val) must be of type int, string given
takesInt(null);  // TypeError: takesInt(): Argument #1 ($val) must be of type int, null given
```

So even without strict types enabled, passing `null` or any other value incompatible with the declared type will throw a `TypeError` at runtime.

> I'd prefer if the validation code was left in. We want to provide good defaults for users.

When native type declarations are in place, the validation code doesn't do anything, since PHP ensures that the arguments have the declared type inside the function. The only way to have that validation code do anything is to remove native type declarations, but I would strongly disagree that this is a good default. It may prevent users from being warned by their IDE that they are passing an incorrect type to the function, leading to runtime errors which could otherwise have been avoided.

Furthermore, if a library doesn't respect a developer's choice of strict or loose type checking, they may have to add extra type casts to function calls, which ironically can *reduce* type safety compared to native type checks. Here's the same example as above, but with manual validation instead of a native type declaration, which caused the developer to manually cast input values:

```php
function takesInt($val) {
    if (!is_int($val)) {
        throw new Exception("Function must be passed an integer");
    }

    var_dump($val);
}

takesInt((int) "5");   // int(5)
takesInt((int) "foo"); // int(0)
takesInt((int) "");    // int(0)
takesInt((int) null);  // int(0)
```

For these reasons among others, it is a best practice to use native type declarations rather than trying to replace them with non-standard manual type checks.

File: .github/workflows/php_ci.yml [hunk 1]
```diff
@@ -1,11 +1,5 @@
 name: PHP CI
-on:
```
Comment (AaronAtDuo) [inline .github/workflows/php_ci.yml]: Any specific reason this is being changed back?

File: src/Client.php [hunk 2]
```diff
@@ -86,15 +84,14 @@ private function getExceptionFromResult($result)
     /**
      * Make HTTPS calls to Duo.
      *
-     * @param string  $endpoint   The endpoint we are trying to hit
-     * @param any     $request    Information to send to Duo
-     * @param boolean $user_agent (Optional)True if we want to send
```
Comment (AaronAtDuo) [inline src/Client.php:91]: Was 'boolean' here inaccurate?

File: .github/workflows/php_ci.yml (hunk 1 above)
Comment (theodorejb) [reply .github/workflows/php_ci.yml]: I changed this so that CI can be run for merge requests from any branch.

File: src/Client.php [hunk 3]
```diff
@@ -161,64 +158,32 @@ private function generateRandomString($state_length)
...
-    private function validateInitialConfig(
-        $client_id,
-        $client_secret,
-        $api_host,
-        $redirect_url,
-        $use_duo_code_attribute
-    ) {
-        if (!is_string($client_id) || !is_string($client_secret) || !is_string($api_host) || !is_string($redirect_url) || !is_bool($use_duo_code_attribute)
-        ) {
-            throw new DuoException(self::PARSING_CONFIG_ERROR);
```
Comment (AaronAtDuo) [inline src/Client.php:185]: Relying on PHP type checking does change the behavior here - what used to be a (vague) DuoException will now be a TypeError if a value of the incorrect type is provided.  I am pondering whether that is a good plan; and if that makes this change backwards incompatible

File: .github/workflows/php_ci.yml [hunk 4]
```diff
@@ -17,10 +11,10 @@ jobs:
 
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
     
       - name: Setup PHP
-        uses: shivammathur/setup-php@86e1ccdd8ddc47bffc29bf667143f363a4cdfdbc
```
Comment (AaronAtDuo) [inline .github/workflows/php_ci.yml:23]: Although we don't do it everywhere like we should, GitHub actually recommends pinning to a SHA for actions; especially third-party actions.  I'm fine with moving to a newer version but we should still pin to a sha here please

File: src/Client.php (hunk 2 above)
Comment (theodorejb) [reply src/Client.php:91]: Yes. `curl_setopt` with the `CURLOPT_USERAGENT` option takes a string (which sets the value of the "User-Agent: " header), not a boolean. See https://curl.se/libcurl/c/CURLOPT_USERAGENT.html and https://www.php.net/manual/en/function.curl-setopt.php.

File: .github/workflows/php_ci.yml (hunk 1 above)
Comment (AaronAtDuo) [reply .github/workflows/php_ci.yml]: We purposely limit to main per our security team's requirements.  As long as the PR is _to_ the main branch, CI should run on it just fine.  So this probably needs to be changed back.

File: src/Client.php (hunk 2 above)
Comment (AaronAtDuo) [reply src/Client.php:91]: past us lied to current us :shame:

File: .github/workflows/php_ci.yml (hunk 4 above)
Comment (theodorejb) [reply .github/workflows/php_ci.yml:23]: Fixed.

File: .github/workflows/php_ci.yml (hunk 1 above)
Comment (theodorejb) [reply .github/workflows/php_ci.yml]: Okay, I changed this back.

File: src/Client.php (hunk 3 above)
Comment (theodorejb) [reply src/Client.php:185]: As a PHP developer the expectation is that a `TypeError` will be thrown for incorrect types - that's how the language and other modern libraries work.

Although virtually any change [could be considered a BC break](https://xkcd.com/1172/), it doesn't make sense why anyone would rely on the specific message thrown when passing a wrong type to these functions. They would just fix their code to pass the correct value.

**Comment by theodorejb** [comment]
@AaronAtDuo I hope this can be merged and a new release tagged, so that I (and others) can enable static analysis on code using the library.

**Comment by AaronAtDuo** [comment]
Yes, sorry, there's been a few distractions recently.  This is on my to-do list to get merged, hopefully in a day or two.

File: src/Client.php (hunk 3 above)
Comment (AaronAtDuo) [reply src/Client.php:185]: Seems reasonable to me.

**Comment by AaronAtDuo** [comment]
@theodorejb For complicated reasons, we don't merge PRs directly in this repo yet.  I have 'merged' the PR by direct push, though.  I'll get the release bumped shortly.  Thanks for the PR, the cleanup, and thanks for using Duo.

**Comment by AaronAtDuo** [comment]
And looks like's it's live on https://packagist.org/packages/duosecurity/duo_universal_php.  Enjoy.

**Comment by theodorejb** [comment]
@AaronAtDuo FYI the Packagist package currently isn't auto-updated (I manually clicked the "Update Now" button there after you pushed the tag to GitHub to make the new release show up).

**Comment by AaronAtDuo** [comment]
Ah that explains it.  We have a spirited internal debate on whether we want auto-updating or not.  Thanks.

**Pull request closed without merging**